trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+        num_users, num_items, num_attributes = validate_user_item_inputs(
+            num_users,
+            num_items,
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            100,
+            1250,
+            10,
+            num_latent_factors,
+        )
+        if model_params is None:
+            model_params = {}
+        num_features_vals = non_none_values(
+            model_params.get("features"), num_latent_factors, num_attributes
+        )
+        if len(num_features_vals) > 1:
+            raise ValueError("Number of latent factors is not the same across inputs.")
+
+        self.num_latent_factors = num_attributes
+        self.model_params = model_params
+        self.als_model = None  # initialize empty model
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])  # empty interactions matrix
+
+        # generate user and item representations as needed
+        if user_representation is None:
+            user_representation = np.zeros((num_users, num_attributes))
+        if item_representation is None:
+            item_representation = Generator(seed=seed).binomial(
+                n=1, p=0.5, size=(num_attributes, num_items)
+            )
+        # if the actual item representation is not specified, we assume
+        # that the recommender system's beliefs about the item attributes
+        # are the same as the "true" item attributes
+        if actual_item_representation is None:
+            actual_item_representation = np.copy(item_representation)
+
+        measurements = [MSEMeasurement()]
+
+        super().__init__(
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            num_users,
+            num_items,
+            num_items_per_iter,
+            probabilistic_recommendations=probabilistic_recommendations,
+            measurements=measurements,
+            verbose=verbose,
+            seed=seed,
+            **kwargs
+        )
+
+    def _update_internal_state(self, interactions):
+        """
+        At each training timestep, we keep track of the user/item interactions.
+        """
+        user_item_ids = tuple(zip(self.users.user_vector, interactions))
+        interaction_df = pd.DataFrame(user_item_ids, columns=["user", "item"])
+        self.all_interactions = self.all_interactions.append(interaction_df, ignore_index=True)
+
+    def update_predicted_scores(self):
+        """
+        Calculates predicted scores for every user-item pair by first running an
+        implicit MF algorithm on the interaction data. A glorified wrapper function,
+        but necessary to keep the functionality in BaseRecommender working.
+        """
+        # if no interactions have happened yet, do a simple dot product
+        if self.all_interactions.size == 0:
```

Comment:
I like that these two functions are separate here. It makes a lot of intuitive sense to me. Maybe for future methods that actually involve some sort of "training" we can follow a similar process!